File: benchmarks/fixtures/webapp_ts/src/validators/payment.ts
/** Payment input validation. */
import { getLogger } from '../utils/helpers';
import { ValidationError } from '../errors';
import { validatePositiveNumber, validateEnum } from './common';

const logger = getLogger("validators.payment");

const CURRENCIES = ["USD", "EUR", "GBP", "JPY", "CAD"];
const METHODS = ["card", "bank_transfer", "wallet"];

/** Validate payment data — name collision with validators/user, api/v1/auth, api/v2/auth. */
export function validate(data: Record<string, unknown>): Record<string, unknown> {
    logger.info("Validating payment data");
    if (!data["amount"]) throw new ValidationError("Amount required", "amount");
    if (!data["currency"]) throw new ValidationError("Currency required", "currency");
    if (!data["user_id"]) throw new ValidationError("User ID required", "user_id");
    const result: Record<string, unknown> = {};
    result["amount"] = validatePositiveNumber(data["amount"], "amount");
    result["currency"] = validateEnum(data["currency"] as string, CURRENCIES, "currency");
    result["user_id"] = data["user_id"];
    result["payment_method"] = data["payment_method"] ? validateEnum(data["payment_method"] as string, METHODS, "payment_method") : "card";
    return result;
}

/** Validate refund data. */
export function validateRefund(data: Record<string, unknown>): Record<string, unknown> {
    if (!data["transaction_id"]) throw new ValidationError("Transaction ID required", "transaction_id");
    const result: Record<string, unknown> = { transaction_id: data["transaction_id"] };
    if (data["amount"]) result["amount"] = validatePositiveNumber(data["amount"], "amount");
    if (data["reason"]) result["reason"] = String(data["reason"]).substring(0, 500);
    return result;
}

File: benchmarks/fixtures/webapp_py/validators/payment.py
"""Payment input validation."""

from typing import Any, Dict

from ..utils.logging import get_logger
from ..exceptions import ValidationError
from .common import validate_positive_number, validate_enum, validate_dict_keys

_logger = get_logger("validators.payment")

SUPPORTED_CURRENCIES = ["USD", "EUR", "GBP", "JPY", "CAD"]
PAYMENT_METHODS = ["card", "bank_transfer", "wallet"]


def validate(data: Dict[str, Any]) -> Dict[str, Any]:
    """Validate payment request data.

    This is the payment-specific validate function (name collision with
    validators.user.validate, api.v1.auth.validate, api.v2.auth.validate).
    """
    _logger.info("Validating payment data")
    validate_dict_keys(data, required=["amount", "currency", "user_id"],
                      optional=["payment_method", "description"])

    result = {}
    result["amount"] = validate_positive_number(data["amount"], "amount")
    result["currency"] = validate_enum(data["currency"], SUPPORTED_CURRENCIES, "currency")
    result["user_id"] = data["user_id"]

    if "payment_method" in data:
        result["payment_method"] = validate_enum(
            data["payment_method"], PAYMENT_METHODS, "payment_method"
        )
    else:
        result["payment_method"] = "card"

    # Validate amount ranges per currency
    max_amounts = {"USD": 999999, "EUR": 999999, "GBP": 999999, "JPY": 99999999, "CAD": 999999}
    max_amount = max_amounts.get(result["currency"], 999999)
    if result["amount"] > max_amount:
        raise ValidationError(
            f"Amount exceeds maximum for {result['currency']}: {max_amount}",
            field="amount",
        )

    return result


def validate_refund(data: Dict[str, Any]) -> Dict[str, Any]:
    """Validate refund request data."""
    validate_dict_keys(data, required=["transaction_id"], optional=["reason", "amount"])

    result = {"transaction_id": data["transaction_id"]}

    if "amount" in data:
        result["amount"] = validate_positive_number(data["amount"], "amount")

    if "reason" in data:
        result["reason"] = str(data["reason"])[:500]

    return result
